### PR TITLE
Update som classifcation

### PR DESCRIPTION
**Pull request by WenzDaniel**
**This PR will change the lineage for > peaklets!**

## What does the code in this PR do / what does it improve?
In this PR we change the SOM plugin such that it only returns a single data-type. It is based on the changes made in https://github.com/XENONnT/straxen/pull/1299. Further, we propagate the SOM and straxen classification, including the SOM subclasses, up to event basics for a better analysis.

I test processed a short run 025423 with the normal offline and SOM context.

### CODE REVIEW

File: straxen/contexts.py
Comment (JYangQi00) [inline straxen/contexts.py:135]: Just for clarification, the PeaksSOM, PeakBasicsSOM, and EventBasicsSOM is so that we do not interfere with the lineage of the non-SOM plugins?

File: straxen/plugins/events/event_basics_som.py
Comment (JYangQi00) [inline straxen/plugins/events/event_basics_som.py:30]: I presume that the `super().compute()` will copy the peak info into EventBasics. But just for clarification, the reason we don't need to use `depends_on = (..., peaks_som)` or something like that is because `peaks` itself is being provided by the PeaksSOM plugin which is registered in the context?

File: straxen/plugins/merged_s2s/merged_s2s.py
Comment (JYangQi00) [inline straxen/plugins/merged_s2s/merged_s2s.py:19]: Presumably changes in this file are all from PR #1299?

File: straxen/plugins/peaklets/peaklet_classification.py
Comment (JYangQi00) [inline straxen/plugins/peaklets/peaklet_classification.py:15]: I have no problems with the change here, but out of curiosity, why did we return peaklet_classification as a dict before? Also, @FaroutYLq , will this be an issue for reprocessing if the version changes but the output is functionally the same?

Comment (FaroutYLq) [reply straxen/plugins/peaklets/peaklet_classification.py:15]: Thanks for the headup @JYangQi00 , very important catch! I would try to avoid touching the standard `peaklet_classification` in this SOM PR. It doesn't look to me that the performance of `peaklet_classification` is meaningfully changed, and if it is just about style we should also have a different PR for this. I would try to avoid bumping version which changes lineage so we have to reprocess again the standard pipeline.

File: straxen/plugins/peaklets/peaklet_classification.py
Comment (FaroutYLq) [inline straxen/plugins/peaklets/peaklet_classification.py:108]: Actually what is the motivation we want to change here?

File: straxen/plugins/peaks/peak_basics.py
Comment (JYangQi00) [inline straxen/plugins/peaks/peak_basics.py:18]: Presumably this is also just from PR #1299 ?

**Review by FaroutYLq — CHANGES_REQUESTED** [review]
I agree with concern by @WenzDaniel and suggest we don't touch the regular `peaklet_classification` in this PR. The rest look good to me, but I have to admit I didn't run tests locally. 
Another general comments: I think for this level of complexity and significance in update we should add some tests. I would let strax* maintainers to comment more here.

File: straxen/plugins/peaks/peaks_som.py
Comment (JYangQi00) [inline straxen/plugins/peaks/peaks_som.py:35]: Perhaps this is a really basic thing that I'm missing, but why are types undefined for merged S2s?

**Review by JYangQi00 — COMMENTED** [review]
Hi Daniel, thanks for the improvements! Everything looks fine to me I am just confused about a few things that I left in the comments. If you think they are fine then I will approve.

File: straxen/contexts.py
Comment (WenzDaniel) [reply straxen/contexts.py:135]: I had a few more thoughts about it during the weekend. This is the idea yes. However, this is only true for not using SOM at all. If we want to use SOM as a cut we also have to use the SOM plugins to do it properly.

File: straxen/plugins/events/event_basics_som.py
Comment (WenzDaniel) [reply straxen/plugins/events/event_basics_som.py:30]: Yes, but the reason is PeakBasicsSOM provides a different version of `peak_basics` which EventBasics(SOM) depends on.

File: straxen/plugins/merged_s2s/merged_s2s.py
Comment (WenzDaniel) [reply straxen/plugins/merged_s2s/merged_s2s.py:19]: Yes, they are. Sorry that it is a bit messy, but I thought you guys would look first at #1299

File: straxen/plugins/peaklets/peaklet_classification.py
Comment (WenzDaniel) [reply straxen/plugins/peaklets/peaklet_classification.py:15]: The change is coming from #1299 and indeed the version bump is purely optional. Since I do not modify the outcome of the plugin we can also merge without version bump if you guys prefer. If you agree I change #1299.

File: straxen/plugins/peaks/peak_basics.py
Comment (WenzDaniel) [reply straxen/plugins/peaks/peak_basics.py:18]: Yes, same as above purely optional.

File: straxen/plugins/peaklets/peaklet_classification.py
Comment (WenzDaniel) [reply straxen/plugins/peaklets/peaklet_classification.py:108]: The motivation is that I use `super.compute()` in PeakletClassifcationSOM and it is easier to work with a returned structure array than a dictionary. This style change has no impact on the remaining straxen workflow.

File: straxen/plugins/peaks/peaks_som.py
Comment (WenzDaniel) [reply straxen/plugins/peaks/peaks_som.py:35]: Because SOM classifies only based on peaklets. Merged S2s is only called on S2s (`type==2`) and we assume that these merged peaks stay S2s (`type==2`). However, for each S2 there are multiple SOM sub-types and it is currently not defined which sub-type should be inherited to the merged S2s after merging. We could do something like the sub-type of the peak with the largest area, but not sure if this make sense. 
If one uses straxen types as classification for peaks then of course `type == straxen_type` and the `straxen_type` of MergedS2s would be defined again, however in that case the information is trivial to retrieve. So I preferred to not add an extra if case and make the code more complicated.

**Comment by WenzDaniel** [comment]
Hmm @FaroutYLq which type of test do you have in mind? I can try to add some in #1299 because there we make actually all technical changes.

File: straxen/contexts.py
Comment (LuisSanchez25) [reply straxen/contexts.py:135]: The lineages of the two will always be separate since the peaklet clasification will have a different liniage as that propagates all the way down the processing pipeline, I thought the main Idea was just to propagate the SOM clusters down the chain no?

Comment (WenzDaniel) [reply straxen/contexts.py:135]: Well yes and no, when I made this PR I made different assumptions. Right now one cannot separate the two.

File: straxen/contexts.py
Comment (WenzDaniel) [inline straxen/contexts.py:22]: These three plugins registered to the default nT context are additional plugins. They will allow us to keep the already processed data as is and provide the additional SOM fields at peak and event level as additional data-types. Thus, after processing on midway we have to use OSG to process `peak_som_classification` and `event_som_classification`.

File: straxen/contexts.py
Comment (WenzDaniel) [inline straxen/contexts.py:135]: These are new plugins which extend the already existing classical versions. With this context we will have a new lineage for everything above peaklets. The same data-type names as for the classical case are used (peaklet_classification, peaks, peak_basics, event_basics), such that we can use the standard processing procedure. 

* `PeakletClassificationSOM` computes first the classical classification before adding the SOM information.
* `PeaksSOM` propagates the information to Peaks such that we can propagate it to peak basics in  `PeakBasicsSOM`. 
* `EventBasicsSOM` is the same as the classical EventBasics, but also extended by the additional SOM fields.

File: straxen/plugins/events/event_basics_som.py
Comment (WenzDaniel) [inline straxen/plugins/events/event_basics_som.py:10]: This plugin is inherited from the classic event-basics and only adds in addition the new SOM fields. It will also provide `event_basics` and is used in the SOM context.

File: straxen/plugins/events/event_basics_som.py
Comment (WenzDaniel) [inline straxen/plugins/events/event_basics_som.py:35]: This plugin is the extra plugin for the default context case in which we provide in addition `event_som_clasification.` For peaks which do not provide any SOM information like S2s -1 is set as default value.

File: straxen/plugins/peaklets/peaklet_classification.py
Comment (WenzDaniel) [inline straxen/plugins/peaklets/peaklet_classification.py:108]: Here I change the plugin return from a dict into an array. There is no impact whatsoever for our processing. Only more convenient to work with when inhering.

File: straxen/plugins/peaklets/peaklet_classification_som.py
Comment (WenzDaniel) [inline straxen/plugins/peaklets/peaklet_classification_som.py:37]: Add option for the future to default back to classic classification if required.

File: straxen/plugins/peaklets/peaklet_classification_som.py
Comment (WenzDaniel) [inline straxen/plugins/peaklets/peaklet_classification_som.py:68]: Inherit method to avoid copy & paste errors.

File: straxen/plugins/peaklets/peaklet_classification_som.py
Comment (WenzDaniel) [inline straxen/plugins/peaklets/peaklet_classification_som.py:70]: In the following are some changes required due to the inheritance and some more explicit naming conventions.

File: straxen/plugins/peaklets/peaklet_classification_som.py
Comment (WenzDaniel) [inline straxen/plugins/peaklets/peaklet_classification_som.py:275]: This plugin is used in the default straxen context as the provided data-type needs to be different from `peaklet_classification`. The rest is inherited from above.

File: straxen/plugins/peaks/peak_basics.py
Comment (WenzDaniel) [inline straxen/plugins/peaks/peak_basics.py:39]: Only moving the dtype in here so it can be better inherited.

File: straxen/plugins/peaks/peak_basics_som.py
Comment (WenzDaniel) [inline straxen/plugins/peaks/peak_basics_som.py:9]: This plugin is used in the SOM context and replaces peak_basics with a new version including the SOM parameters.

File: straxen/plugins/peaks/peaks_som.py
Comment (WenzDaniel) [inline straxen/plugins/peaks/peaks_som.py:9]: This plugin is used in the SOM context to propagate the SOM parameters from `peaklet_classification` to `peaks`.

File: straxen/plugins/peaks/peaks_som.py
Comment (WenzDaniel) [inline straxen/plugins/peaks/peaks_som.py:44]: This is again a stand alone plugin to propagate the SOM settings to peaks for the default nT context using the default classification.

**Review by WenzDaniel — COMMENTED** [review]
Hey guys I think I made all needed modifications. I also added some comments to the code for guidance. I conducted the following test to see if everything works as intended:

1. Process run 025423 up to event basics using the current master using the default offline context. Switch to this branch and see if I can still load peaklet_classifcation, merged_s2s, peak_basics and event_basics with the default nT offline context.
2. Processed for the same run the three new SOM data-types and checked if I can load them with meaningful values too.
3. I repeated the test of 1. with run `053494` using the XENONnT online context.

If you guys do not have any objections I think we can merge.

**Review by FaroutYLq — APPROVED** [review]

**Comment by JYangQi00** [comment]
I checked that the That peaklet_classification (SOM context) has the same straxen_type as the type of peaklet_classification (nT online context), and the same som fields as peaklet_classification_som (also in nT context). I did similar things as well for the peak and event basics as well, but only looking at the S1 SOM information (as intended by this PR). Everything looks good on my end and I think we are good to merge!

**Pull request merged**
